src/Synthesiser.cpp

```diff
@@ -829,6 +829,219 @@ void Synthesiser::emitCode(std::ostream& out, const RamStatement& stmt) {
+                out << preamble.str();
+                visitTupleOperation(aggregate, out);
+                PRINT_END_COMMENT(out);
+                return;
+            }
+
+            out << "bool shouldRunNested = false;\n";
+
+            // init result and reduction operation
+            std::string init;
+            switch (aggregate.getFunction()) {
+                case AggregateOp::MIN: init = "MAX_RAM_SIGNED"; break;
+                case AggregateOp::FMIN: init = "MAX_RAM_FLOAT"; break;
+                case AggregateOp::UMIN: init = "MAX_RAM_UNSIGNED"; break;
+                case AggregateOp::MAX: init = "MIN_RAM_SIGNED"; break;
+                case AggregateOp::FMAX: init = "MIN_RAM_FLOAT"; break;
+                case AggregateOp::UMAX: init = "MIN_RAM_UNSIGNED"; break;
+                case AggregateOp::COUNT:
+                    init = "0";
+                    out << "shouldRunNested = true;\n";
+                    break;
+                case AggregateOp::MEAN: init = "0"; break;
+                case AggregateOp::FSUM:
+                case AggregateOp::USUM:
+                case AggregateOp::SUM:
+                    init = "0";
+                    out << "shouldRunNested = true;\n";
+                    break;
+            }
+
+            // Set reduction operation
+            std::string op;
+            switch (aggregate.getFunction()) {
+                case AggregateOp::MIN:
+                case AggregateOp::FMIN:
+                case AggregateOp::UMIN: {
+                    op = "min";
+                    break;
+                }
+
+                case AggregateOp::MAX:
+                case AggregateOp::FMAX:
+                case AggregateOp::UMAX: {
+                    op = "max";
+                    break;
+                }
+
+                case AggregateOp::MEAN:
+                case AggregateOp::FSUM:
+                case AggregateOp::USUM:
+                case AggregateOp::COUNT:
+
+                case AggregateOp::SUM: {
+                    op = "+";
+                    break;
+                }
+                default: fatal("Unhandled aggregate operation");
+            }
+            // res0 stores the aggregate result
+            std::string sharedVariable = "res0";
+
+            std::string type;
+            switch (getTypeAttributeAggregate(aggregate.getFunction())) {
+                case TypeAttribute::Signed: type = "RamSigned"; break;
+                case TypeAttribute::Unsigned: type = "RamUnsigned"; break;
+                case TypeAttribute::Float: type = "RamFloat"; break;
+
+                case TypeAttribute::Symbol:
+                case TypeAttribute::Record: type = "RamDomain"; break;
+            }
+            out << type << " res0 = " << init << ";\n";
+            if (aggregate.getFunction() == AggregateOp::MEAN) {
+                out << "RamUnsigned res1 = 0;\n";
+                sharedVariable += ", res1";
+            }
```

Comment:
`res0` needs to be RamFloat when the operation is a mean. Otherwise integer overflow can easily happen.
```suggestion
            if (aggregate.getFunction() == AggregateOp::MEAN) {
                out << "RamFloat res0 = " << init << ";\n";
                out << "RamUnsigned res1 = 0;\n";
                sharedVariable += ", res1";
            } else {
                out << type << " res0 = " << init << ";\n";
            }
```

Reply:
Is it necessary though? all mean functions return RamFloat for their type attribute (getTypeAttributeAggregate)

Reply:
True, but we're not just using res0 as the return type, but as part of the calculation.

If you're confident this is safe, then that's fine. If there's not already such a test, then it would be good if one of the aggregates took the mean of two large ints such that an overflow occurs if the maths is incorrect. That would check that this aggregate keeps working correctly.

Reply:
Okay got you I will add something like that